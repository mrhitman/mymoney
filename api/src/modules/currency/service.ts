import axios from 'axios';
import xml2js from 'xml2js';
import redis from '../../services/redis';

class CurrencyProvider {
  async latest(base?: string) {
    const client = redis;
    const rates = await client.get('rates');
    if (rates) {
      return JSON.parse(rates);
    }
    const response = await fixer.latest({ base });
    await client.set('rates', JSON.stringify(response), 'EX', 3600);
    return response;
  }

  // ISO_4217 standart
  async getIsoInfo() {
    const url = 'https://www.currency-iso.org/dam/downloads/lists/list_one.xml';
    const body = await axios.get(url);
    return new Promise((resolve, reject) => {
      xml2js.parseString(body.data, { explicitArray: false }, (err, result) => {
        if (err) {
          return reject(err);
        }
        resolve(result);
      });
    });
  }
}
