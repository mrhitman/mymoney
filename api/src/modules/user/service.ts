import { hash } from 'bcrypt';
import User from 'src/models/user';
import { BadRequest, NotFound } from 'ts-httpexceptions';
import CreateUserDto from './dto/CreateUserDto';
import UpdateUserDto from './dto/UpdateUserDto';
import { userInfo } from 'os';

export class UserProvider {
  public async create(dto: CreateUserDto) {
    if (await User.query().findOne({ email: dto.email })) {
      throw new BadRequest('User already exists');
    }

    return User.query().insert({
      email: dto.email,
      first_name: dto.first_name,
      middle_name: dto.middle_name,
      last_name: dto.last_name,
      password: await hash(dto.password, 10),
      last_sync: new Date(),
    });
  }

  public async update(dto: UpdateUserDto) {
    const user = await User.query().findOne({ email: dto.email });

    if (!user) {
      throw new NotFound('No such user found');
    }

    await user.$query().update({ last_sync: new Date() });
    return user;
  }

  public async get() {
    return User.query();
  }
}

export default UserProvider;
