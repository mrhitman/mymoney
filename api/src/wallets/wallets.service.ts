import { Injectable } from '@nestjs/common';
import User from 'src/database/models/user.model';
import Wallet from 'src/database/models/wallet.model';
import { v4 as uuid } from 'uuid';
import Transaction from 'src/database/models/transaction.model';

@Injectable()
export class WalletsService {
  public async getAll(user: User) {
    return Wallet.query().where({ userId: user.id });
  }

  public async getWallet(id: string) {
    return Wallet.query().findById(id);
  }

  public async performOperation(wallet: Wallet, trx: Transaction) {
    const pocket = wallet.pockets.find(
      (p) => p.currencyId === trx.currencyId,
    ) || {
      id: uuid(),
      currencyId: trx.currencyId,
      amount: 0,
    };

    pocket.amount += trx.amount * (trx.type === 'income' ? 1 : -1);
  }
}
