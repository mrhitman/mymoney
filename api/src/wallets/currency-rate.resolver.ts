import { Float, Parent, ResolveProperty, Resolver } from '@nestjs/graphql';
import { CurrenciesService } from 'src/currencies/currencies.service';
import { CurrencyDto } from 'src/currencies/dto/currency.dto';

@Resolver((of) => CurrencyDto)
export class CurrencyRateResolver {
  constructor(protected service: CurrenciesService) {}

  @ResolveProperty('rate', () => Float)
  async getRate(@Parent() currency: CurrencyDto) {
    const rates = await this.service.rates();
    return rates.rates[currency.name] || 0;
  }
}
