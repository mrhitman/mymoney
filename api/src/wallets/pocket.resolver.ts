import { Parent, ResolveField, Resolver } from '@nestjs/graphql';
import { CurrencyDto } from 'src/currencies/dto/currency.dto';
import { DataLoader } from '../dataloader';
import { PocketDto } from './dto/pocket.dto';

@Resolver((of) => PocketDto)
export class PocketResolver {
  constructor(private readonly loader: DataLoader) {}

  @ResolveField('currency', () => CurrencyDto)
  async getCurrency(@Parent() pocket: PocketDto) {
    return this.loader.currency.load(pocket.currencyId);
  }
}
