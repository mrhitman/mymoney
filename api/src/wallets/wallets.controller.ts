import { Controller, Get, Request, UseGuards } from '@nestjs/common';
import { ApiHeader } from '@nestjs/swagger';
import { JwtAuthGuard } from 'src/auth/guards/jwt-auth.guard';
import { WalletsService } from './wallets.service';

@Controller('wallets')
@UseGuards(JwtAuthGuard)
@ApiHeader({ name: 'jwt token' })
export class WalletsController {
  constructor(private readonly service: WalletsService) {}

  @Get()
  public async getAll(@Request() req) {
    return this.service.getAll(req.user);
  }
}
