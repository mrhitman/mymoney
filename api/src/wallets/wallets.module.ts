import { CacheModule, Module } from '@nestjs/common';
import redisStore from 'cache-manager-redis-store';
import { CurrenciesModule } from 'src/currencies/currencies.module';
import { DataLoader } from 'src/dataloader';
import { CurrencyRateResolver } from './currency-rate.resolver';
import { PocketResolver } from './pocket.resolver';
import { WalletsResolver } from './wallets.resolver';
import { WalletsService } from './wallets.service';

@Module({
  imports: [
    CurrenciesModule,
    CacheModule.register({
      store: redisStore,
      url: process.env.REDIS_URL,
    }),
  ],
  providers: [
    WalletsResolver,
    PocketResolver,
    CurrencyRateResolver,
    WalletsService,
    DataLoader,
  ],
})
export class WalletsModule { }
