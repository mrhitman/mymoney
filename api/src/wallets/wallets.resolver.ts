import { UseGuards } from '@nestjs/common';
import { Args, Mutation, Query, Resolver } from '@nestjs/graphql';
import { CurrentUser } from 'src/auth/current-user';
import { GqlAuthGuard } from 'src/auth/guards/gql-auth.quard';
import User from 'src/database/models/user.model';
import { WalletDto } from './dto/wallet.dto';
import { WalletCreate } from './input/wallet-create';
import { WalletUpdate } from './input/wallet-update';
import { WalletsService } from './wallets.service';

@Resolver((of) => WalletDto)
export class WalletsResolver {
  constructor(private readonly service: WalletsService) {}

  @UseGuards(GqlAuthGuard)
  @Query((returns) => [WalletDto])
  async wallets(@CurrentUser() user: User) {
    return this.service.getAll(user);
  }

  @UseGuards(GqlAuthGuard)
  @Query((returns) => WalletDto)
  async wallet(@CurrentUser() user: User, @Args('id') id: string) {
    return this.service.findOne(user, id);
  }

  @UseGuards(GqlAuthGuard)
  @Mutation((returns) => WalletDto)
  async createWallet(
    @CurrentUser() user: User,
    @Args('walletCreateData')
    data: WalletCreate,
  ) {
    return this.service.create(user, data);
  }

  @UseGuards(GqlAuthGuard)
  @Mutation((returns) => WalletDto)
  async updateWallet(
    @CurrentUser() user: User,
    @Args('walletUpdateData')
    data: WalletUpdate,
  ) {
    return this.service.update(user, data);
  }

  @UseGuards(GqlAuthGuard)
  @Mutation((returns) => WalletDto)
  async deleteWallet(@CurrentUser() user: User, @Args('id') id: string) {
    return this.service.delete(user, id);
  }
}
