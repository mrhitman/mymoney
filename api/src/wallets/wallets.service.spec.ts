import { Test, TestingModule } from '@nestjs/testing';
import User from 'src/database/models/user.model';
import { WalletsService } from './wallets.service';
import { Chance } from 'chance';
import Wallet from 'src/database/models/wallet.model';

describe('WalletsService', () => {
  let service: WalletsService;
  let user: User;
  const chance = new Chance(42);
  const userData = {
    id: 1,
    firstName: chance.name(),
    email: chance.email(),
    password: chance.word({ length: 10 }),
    imageUrl: '',
    additional: {},
  };

  beforeEach(async () => {
    jest.clearAllMocks();
    (<jest.Mock>User.query).mockImplementation(() => ({
      where: jest.fn().mockReturnThis(),
      first: jest.fn().mockReturnValue(userData),
      findById: jest.fn().mockImplementation((id) => {
        return {
          '1': { ...userData },
        }[id.toString()];
      }),
    }));
    (<jest.Mock>Wallet.query).mockImplementation(() => ({
      where: jest.fn().mockReturnThis(),
      whereNot: jest.fn().mockReturnValue([
        {
          id: 1,
          name: 'test_wallet',
          userId: 1,
          description: 'test_wallet_desc',
          pockets: [],
        },
      ]),
    }));
    user = await User.query().findById(1);
    const module: TestingModule = await Test.createTestingModule({
      providers: [WalletsService],
    }).compile();

    service = module.get<WalletsService>(WalletsService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('get all', () => {
    it(' success', async () => {
      const wallets = await service.getAll(user);
      expect(wallets).toMatchSnapshot();
    });
  });
});
