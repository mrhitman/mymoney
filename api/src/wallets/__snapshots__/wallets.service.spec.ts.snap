// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletsService get all  success 1`] = `
Array [
  Object {
    "description": "test_wallet_desc",
    "id": 1,
    "name": "test_wallet",
    "pockets": Array [],
    "userId": 1,
  },
]
`;
